replica/transition.go

```diff
@@ -34,24 +70,154 @@ type Proposed struct {
 type PreVoted struct {
 	block.SignedPreVote
 }
 
-// IsTransition implements the `Transition` interface for the `PreVoted` event.
+// IsTransition implements the `Transition` interface for the
+// `PreVoted` event.
 func (preVoted PreVoted) IsTransition() {
 }
 
-// A PreCommitted polka has been signed and broadcast by another `Replica`.
+// A PreCommitted polka has been signed and broadcast by another
+// `Replica`.
 type PreCommitted struct {
 	block.SignedPreCommit
 }
 
-// IsTransition implements the `Transition` interface for the `PreCommitted` event.
+//TODO: should we have Commited as well?
+
+// IsTransition implements the `Transition` interface for the
+// `PreCommitted` event.
 func (preCommitted PreCommitted) IsTransition() {
 }
+
+func (buffer *transitionBuffer) Enqueue(transition Transition) {
+	switch transition := transition.(type) {
+	case Proposed:
+		buffer.initMapKey(transition.Height)
+		queue := buffer.buf[transition.Height]
+		if tran, ok := queue.peek(); ok {
+			switch tran.(type) {
+			case Proposed:
+			case PreVoted:
+				panic("You Enqueued a PreVoted before a Proposed")
+			case PreCommitted:
+				panic("You Enqueued a PreCommitted before a Proposed")
+			default:
+			}
+		}
+		queue.enqueue(transition)
+	case PreVoted:
+		buffer.initMapKey(transition.Height)
+		queue := buffer.buf[transition.Height]
+		if tran, ok := queue.peek(); ok {
+			switch tran.(type) {
+			case Proposed:
+				queue.reset()
+			case PreVoted:
+			case PreCommitted:
+				panic("You Enqueued a PreCommitted before a PreVoted")
+			default:
+			}
+		}
+		queue.enqueue(transition)
+	case PreCommitted:
+		buffer.initMapKey(transition.Polka.Height)
+		queue := buffer.buf[transition.Polka.Height]
+		if tran, ok := queue.peek(); ok {
+			switch tran.(type) {
+			case Proposed:
+				queue.reset()
+			case PreVoted:
+				queue.reset()
+			case PreCommitted:
+			default:
+			}
+		}
+		queue.enqueue(transition)
+	default:
+		buffer.immediate.enqueue(transition)
+	}
+}
+
+func (buffer *transitionBuffer) Dequeue(height block.Height) (Transition, bool) {
+	if tran, ok := buffer.immediate.dequeue(); ok {
+		return tran, true
+	}
+	if queue, ok := buffer.buf[height]; ok {
+		return queue.dequeue()
+	}
+	return nil, false
+}
+
+func (buffer *transitionBuffer) Drop(height block.Height) {
+	for k := range buffer.buf {
+		if k < height {
+			delete(buffer.buf, k)
+		}
+	}
+}
+
+// Convenience function to make sure the map already has a Queue
+// for the provided height
+func (buffer *transitionBuffer) initMapKey(height block.Height) {
+	if _, ok := buffer.buf[height]; !ok {
+		buffer.buf[height] = newQueue()
+	}
+}
+
+// The logic behind the buf is to delete the transitionQueue whenever
+// we get a Transition that makes the previous messages obsolete
+type transitionBuffer struct {
+	buf       map[block.Height]*transitionQueue
+	immediate *transitionQueue
+}
+
+func newQueue() *transitionQueue {
+	return &transitionQueue{
+		queue: make([]Transition, 0),
```

Comment:
We can initialise this queue with a capacity. We should have some vague idea about the bounds of this transition which allow us to avoid some runtime allocations. For example, when there are 100 nodes participating in the consensus algorithm, we can make a reasonable estimate that we will usually not need to queue more than 67 transitions for any specific height.

Reply:
Added